Examples/Algorithms/Geant4/src/Geant4Simulation.cpp

```diff
@@ -9,151 +9,92 @@
+    : IAlgorithm(std::move(name), level) {
+  if (cfg.inputParticles.empty()) {
     throw std::invalid_argument("Missing input particle collection");
   }
-
-  // If we are in VERBOSE mode, set the verbose level in Geant4 to 2.
-  // 3 would be also possible, but that produces infinite amount of output.
-  const int geantVerboseLevel =
-      logger().level() == Acts::Logging::VERBOSE ? 2 : 0;
-  m_cfg.runManager->SetVerboseLevel(geantVerboseLevel);
-  G4EventManager::GetEventManager()->SetVerboseLevel(geantVerboseLevel);
-  G4EventManager::GetEventManager()->GetTrackingManager()->SetVerboseLevel(
-      geantVerboseLevel);
-  G4EventManager::GetEventManager()->GetStackManager()->SetVerboseLevel(
-      geantVerboseLevel);
-
-  // Suppress the printing of physics information.
-  if (logger().level() > Acts::Logging::DEBUG) {
-#if G4VERSION_NUMBER >= 1100
-    G4HadronicParameters::Instance()->SetVerboseLevel(geantVerboseLevel);
-    G4HadronicProcessStore::Instance()->SetVerbose(geantVerboseLevel);
-    G4EmParameters::Instance()->SetIsPrintedFlag(true);
-#endif
+  if (!cfg.detectorConstructionFactory) {
+    throw std::invalid_argument("Missing detector construction factory");
+  }
+  if (!cfg.randomNumbers) {
+    throw std::invalid_argument("Missing random numbers");
   }
 
-  // Set the detector construction
-  m_cfg.runManager->SetUserInitialization(m_cfg.detectorConstruction);
-
-  // Set the primary generator action
-  m_cfg.runManager->SetUserAction(m_cfg.primaryGeneratorAction);
+  m_logger = Acts::getDefaultLogger("Geant4", level);
 
-  // Set the configured user actions
-  m_cfg.runManager->SetUserAction(m_cfg.runAction);
-  m_cfg.runManager->SetUserAction(m_cfg.eventAction);
-  m_cfg.runManager->SetUserAction(m_cfg.trackingAction);
-  m_cfg.runManager->SetUserAction(m_cfg.steppingAction);
+  m_eventStore = std::make_shared<EventStore>();
 
-  // Initialize the Geant4 run manager
-  m_cfg.runManager->Initialize();
+  // tweek logging
+  {
+    // If we are in VERBOSE mode, set the verbose level in Geant4 to 2.
+    // 3 would be also possible, but that produces infinite amount of output.
+    m_geant4Level = logger().level() == Acts::Logging::VERBOSE ? 2 : 0;
+  }
+}
 
-  // Please note:
-  // The following two blocks rely on the fact that the Acts
-  // detector constructions cache the world volume
+ActsExamples::Geant4SimulationBase::~Geant4SimulationBase() = default;
 
-  // Set the magnetic field
-  if (m_cfg.magneticField != nullptr) {
-    ACTS_INFO("Setting ACTS configured field to Geant4.");
-    // Get the g4World cache
-    G4VPhysicalVolume* g4World = m_cfg.detectorConstruction->Construct();
-    /// Set the field ot the G4Field manager
-    G4FieldManager* fieldMgr = new G4FieldManager();
-    fieldMgr->SetDetectorField(m_cfg.magneticField);
-    fieldMgr->CreateChordFinder(m_cfg.magneticField);
+void ActsExamples::Geant4SimulationBase::commonInitialization() {
```

Comment:
Is there a  reason why this is not done in the Constructor of `Geant4SimulationBase`?

Reply:
the problem is that we need a run manager for this and it is coupled to the physics list now which depends on sim or material recording. I would also like to use the constructor for this but I dont think this is possible right now